qutip/core/data/properties.pyx

```diff
@@ -135,6 +135,8 @@ cpdef bint isherm_csr(CSR matrix, double tol=-1):
                 # Pointer into the "transposed" matrix.
                 ptr_t = out_row_index[col]
                 out_row_index[col] += 1
+                if nrows != matrix.row_index[ptr_t]:
+                    return _isherm_csr_full(matrix, tol)
```

Comment:
```suggestion
                if row != matrix.col_index[ptr_t]:
                    return False
```
The code as it was in the PR seems wonky -- `.row_index[...]` should be indexed by a row number, not by a ptr, and wouldn't always be `nrows` in any case, so the code falls back to `_isherm_csr_full` almost always (even when the matrix is hermitian).

The suggested code checks that the column of the `ptr_t` is the correct one and thus avoids having to fall back to `_isherm_csr_full`.

This logic is quite tricky, so please check mine too!

Reply:
Yes this test was wrong, it would always call `_isherm_csr_full`. Your test work, thank you for catching. But we must still fall back to `_isherm_csr_full` since we can have a number under the `tol` on one side and nothing on the other.